zilencer/views.py

```diff
@@ -1749,3 +1759,269 @@ def remote_server_check_analytics(request: HttpRequest, server: RemoteZulipServe
+            "push_account_id": remote_push_device.push_account_id,
+        }
+        if remote_push_device.token_kind == RemotePushDevice.TokenKind.APNS:
+            apns_message_payload = {
+                **apns_base_message_payload,
+                **message_payload,
+            }
+            apns_requests.append(
+                aioapns.NotificationRequest(
+                    apns_topic=remote_push_device.ios_app_id,
+                    device_token=remote_push_device.token,
+                    message=apns_message_payload,
+                    time_to_live=24 * 3600,
+                    # TODO: priority
+                )
+            )
+            apns_remote_push_devices.append(remote_push_device)
+        else:
+            fcm_requests.append(
+                firebase_messaging.Message(
+                    data=message_payload,
+                    token=remote_push_device.token,
+                    android=firebase_messaging.AndroidConfig(priority=priority),
+                )
+            )
+            fcm_remote_push_devices.append(remote_push_device)
+
+    if len(apns_requests) > 0:
+        apple_successfully_sent_count = send_e2ee_push_notification_apple(
+            apns_requests, apns_remote_push_devices, delete_device_ids
+        )
+
+    if len(fcm_requests) > 0:
+        android_successfully_sent_count = send_e2ee_push_notification_android(
+            fcm_requests, fcm_remote_push_devices, delete_device_ids
+        )
+
+    return {
+        "apple_successfully_sent_count": apple_successfully_sent_count,
+        "android_successfully_sent_count": android_successfully_sent_count,
+        "delete_device_ids": delete_device_ids,
+    }
+
+
+class SendE2EEPushNotificationPayload(BaseModel):
+    realm_uuid: str
+    device_id_to_encrypted_data: dict[str, str]
+
+
+@typed_endpoint
+def remote_server_send_e2ee_push_notification(
+    request: HttpRequest,
+    server: RemoteZulipServer,
+    *,
+    payload: JsonBodyPayload[SendE2EEPushNotificationPayload],
+) -> HttpResponse:
+    from corporate.lib.stripe import get_push_status_for_remote_request
+
+    remote_realm = get_remote_realm_helper(request, server, payload.realm_uuid)
+    if remote_realm is None:
+        raise MissingRemoteRealmError
+    else:
+        remote_realm.last_request_datetime = timezone_now()
+        remote_realm.save(update_fields=["last_request_datetime"])
+
+    push_status = get_push_status_for_remote_request(server, remote_realm)
+    log_data = RequestNotes.get_notes(request).log_data
+    assert log_data is not None
+    log_data["extra"] = f"[can_push={push_status.can_push}/{push_status.message}]"
+    if not push_status.can_push:
+        reason = push_status.message
+        raise PushNotificationsDisallowedError(reason=reason)
+
+    # TODO: Log 'remote_queue_latency'.
+    # We need `message.sent_time` to log the latency. But that'd be
+    # a private data which shouldn't be accessible to bouncer, right ?
```

Comment:
I don't think this is something that we need to log for bounced requests. To the extent that logging latency metadata is useful:
- The server could be doing end-to-end logging with latency. It has the send timestamp, and the time it receives the HTTP 200 for the bounce request could reasonably be when it reports.
- The bouncer can do latency for how long it took to process the request. I'm not sure there's actually anything to do there; the `200` line in the bouncer server logs has the request processing time, which is basically this.

I think that's the same conclusion reached in the chat.zulip.org thread, so I think we can delete this comment and maybe open an issue if that thread ends up suggesting it.